### ISSUE TITLE
Gaussian Noise Model

### ISSUE THREAD

**Issue by jbgao** [cabralpinto/modular-diffusion#12]
Not sure why the definition def approximate(self, z: Tensor, t: Tensor, hat: Tensor) in Gaussian noise does not consider two different parameters x and epsilon.  The forged approximate distribution is the case with epsilon parameter.

Could you please have a look?

**Comment by cabralpinto** [comment]
Hi @jbgao! Thanks for your issue. I'm sorry I didn't get to it sooner :). In Modular Diffusion, the `hat` Tensor is the parameter being predicted. This could be $x$ or $\epsilon$ or something else, depending how the rest of your model is setup. For instance, if you want to predict $\epsilon$ set the `parameter` argument of the `Gaussian` noise module to `epsilon` and it will do the necessary calculations. Also, if you're using the `Simple` loss function, be sure to also set its `parameter` to `epsilon` so that everything works.

**Comment by jbgao** [comment]
Thank you very much for the answer.  That was clear regarding using either noise or data approximation network.  

However, my question was about sampling method. The sampling distribution (aka p_{theta}) defined in approximate function is for the case of epsilon-parameter, that is  eqn (125) in https://arxiv.org/pdf/2208.11970.pdf.  Or it was mentioned in Noise Type section in https://cabralpinto.github.io/modular-diffusion/guides/custom-modules/.

However in the case of x-parameter, the network hat is the approximation to the data x_0, that is \hat{x}(x_t, t) \approx x_0.  The approximated distribution p_{theta} should be eqn (94) in  https://arxiv.org/pdf/2208.11970.pdf.    Now sure whether this is right.

Thank you

**Comment by cabralpinto** [comment]
Hi @jbgao, can you please point to the piece code that is giving you trouble? That might help me assist you better. By the way, I will be afk for the next 9 hours. Good luck!

**Comment by jbgao** [comment]
Thanks again.  Here is the screen shots of noise.py
![image](https://github.com/cabralpinto/modular-diffusion/assets/4993409/e47f2185-9ce5-406e-86dd-f2d60cf228fe)

As you can see the return N is the one with mean in (125) in https://arxiv.org/pdf/2208.11970.pdf.    This is the case for epsilon network (hat)

There should be a if statement for self.parameter = 'x', such as according to (94)  in https://arxiv.org/pdf/2208.11970.pdf, to define the mean of N as 
self.q3[t] * z + self.q4[t] * hat[0]

or 
if self.parameter == "x":
    N(self.q3[t] * z + self.q4[t] * hat[0],
            self.q5[t]
            if self.variance == "fixed"
            else torch.exp(hat[1] * self.p3[t] + (1 - hat[1]) * self.p4[t])
            if self.variance == "range"
            else hat[1],
        )

**Comment by cabralpinto** [comment]
If you take a look at the `schedule` method, you'll see that `self.p1` and `self.p2` are conditioned based on `self.parameter`:

```python
def schedule(self, alpha: Tensor) -> None:
    ...
    if self.parameter == "x":
        self.p1, self.p2 = self.q3, self.q4
    elif self.parameter == "epsilon":
        self.p1 = 1 / alpha.sqrt()
        self.p2 = (alpha - 1) / ((1 - delta).sqrt() * alpha.sqrt())
```

So `self.p1[t] * z + self.p2[t] * hat[0]` does equal `self.q3[t] * z + self.q4[t] * hat[0]` when `self.parameter == "x"`. Does this make sense?

PS: Thanks for bringing that commented code to my attention. I was debugging the library before the release and forgot to delete them. I have now [erased it](https://github.com/cabralpinto/modular-diffusion/commit/0febe3d04a56f849255b2ec15fcc3020f36a96b2).

**Comment by jbgao** [comment]
Thank you for this.  I overlooked that part.    Now this is clear to me. :-)

**Comment by cabralpinto** [comment]
No worries man. In the future, if you have any more questions, contact me on our [Discord](https://discord.gg/pHVpvwXH) where I usually respond faster.
